### DISCUSSION TITLE
Async await implementation

### DISCUSSION THREAD

**Discussion by patskovn** [groue/GRDB.swift#1065] [Ideas]
New swift concurrency model consists from several parts. Two most valuable of them is synchronous-like execution flow (async await itself) and performance and safety improvements (prevention of priority inversion, much less context switches between threads via cooperative thread pool). So for proper adoption of whole swift concurrency idea we should consider several phases. Let's just describe them:
* _Phase 1_. Providing quick async await wrapper via continuations. 
* _Phase 2_. Increasing performance, illuminating extra context switches and usage of cooperative thread pools through `actor` synchronisation mechanism under the hood for `AsyncDatabaseQueue`, separating async version of protocols from sync version.
* _Phase 3_. Usage of custom executors for `AsyncDatabasePool`. Phase 3 is currently not implementable because of lack of [custom executors support](https://github.com/rjmccall/swift-evolution/blob/custom-executors/proposals/0000-custom-executors.md).

**This discussion will be just about _Phase 1_**. I will provide drafts, difficulties and possibilities for adoption. A little bit later I will create discussion for _Phase 2_ with more options and reasoning to show idea and take more feedback and thoughts from @groue and other folks. 
Let's code!

Implementation of async await via continuation if pretty straightforward for any method. Just:
```swift
func asyncAwaitVersionOfMethod<T>(_ value: @escaping (Database) throws -> T) async throws -> T {
    try await withUnsafeThrowingContinuation { continuation in // Wrap with continuation
        asyncVersionOfMethodWithClosure { result in // Call method with closure completion
            do {
	        try continuation.resume(returning: value(result.get()))
            } catch {
	        continuation.resume(throwing: error)
            }
        }
    }
}
```

We have a problems with [incorrect implementation of method overloading diffing just with async keyword](https://bugs.swift.org/browse/SR-15150). All problems described below happened only with protocols. When we use just concrete classes, everything works perfectly. I tried several approaches to workaround the issue and here is my notes:

<details>
  <summary>1. The case from swift bug that is not working</summary>
  We can't create tho methods in protocols diffing with just async keyword. Compilation fails.

```swift
protocol TestProtocol {
    func test()
    func test() async
}

class TestClass: TestProtocol {
    func test() {}
    func test() async {}
}

```
</details>

<details>
  <summary>2. Workaround that works for methods with any parameters</summary>
  But! We can just create protocol extension that wraps one of another protocol methods. This compiles successfully.

  ```swift
  protocol TestProtocol {
      func test(closure: () -> Int) -> Int
  }
  extension TestProtocol {
      func test(closure: () -> Int) async -> Int {
          return closure()
      }
  }
  
  class TestClass: TestProtocol {
      func test(closure: () -> Int) -> Int { return closure() } 
  }

  class Consumer {
      let t = TestClass()
      func caller() async -> Int {
            return await t.test(closure: { return 0 })
      }
  }
  ```
</details>

<details>
  <summary>3. But this not really work when we have generics</summary>
  This still compiles without errors. 

**But!** We are facing problems from the call side when we start using generics. Compiler shows us warning`No 'async' operations occur within 'await' expression` when we call `await t.test(closure: { return 0 })`
  
  ```swift
  protocol TestProtocol {
      func test<T>(closure: () -> T) -> T
  }
  extension TestProtocol {
      func test<T>(closure: () -> T) async -> T {
          return closure()
      }
  }
  
  class TestClass: TestProtocol {
      func test<T>(closure: () -> T) -> T { return closure() }
  }

  class Consumer {
      let t = TestClass()
      func caller() async -> Int {
            return await t.test(closure: { return 0 })
      }
  }
  ```

</details>

<details>
  <summary>4. Lets apply next filthy workaround which looks like solves our problems again</summary>

  Bring good old `@_disfavoredOverload`. This compiles without any warnings and errors.
  
  ```swift
  protocol TestProtocol {
     // NOTE: We are using @_disfavoredOverload on the sync method in protocol to handle method resolving using protocol
      @_disfavoredOverload
      func test<T>(closure: () -> T) -> T
  }
  extension TestProtocol {
      func test<T>(closure: () -> T) async -> T {
          return closure()
      }
  }
  
  class TestClass: TestProtocol {
     // NOTE: We are using @_disfavoredOverload on the sync method in protocol to handle method resolving using class
      @_disfavoredOverload
      func test<T>(closure: () -> T) -> T { return closure() }
  }

  class Consumer {
      let t = TestClass()
      func caller() async -> Int {
            return await t.test(closure: { return 0 })
      }

      func protocolCaller() async -> Int {
          let ptorocolVersion: TestProtocol = t
          return await t.test(closure: { return 0 })
      }

      func syncVersionCaller() -> Int {
            return t.test(closure: { return 0 })
      }
  }
  ```

</details>

P.S. Everything tested with 
Xcode Version 13.0 (13A233)
swift-driver version: 1.26.9 Apple Swift version 5.5 (swiftlang-1300.0.29.102 clang-1300.0.28.1)
Target: x86_64-apple-macosx12.0

**Comment by patskovn** [comment]
Hello, @groue 
What are your thoughts about Phase 1?
Can I implement it and make a PR? Or should I at first write more about Phase 2 and actors? Or, maybe, you will try to investigate another workarounds to avoid usage of private `@_disfavoredOverload`?

**Reply by groue** [reply]
Hello @Nekitosss,

Thanks for starting this discussion! I have great hopes that the library eventually ships with great async apis 👍 

First, do we agree that Phase 1 could be implemented as extension methods in an application? I'm not asking if it would be relevant or not to ship those methods in the library, but if it is mandatory to ship them in the lib because users can not implement them on their own.

Next, `AsyncDatabaseQueue` and `AsyncDatabasePool`, if they are actor-based, will only provide asynchronous database access. We could expose `nonisolated` functions that can be called synchronously, but we would then have to implement isolation ourselves, and all benefits of actors would be lost. So I assume that actor-based database apis can *only* be async. And this is concerning, as explained below.

I agree it's super cool to have async access to the database. But I do not agree that sync access is uncool. The cardinal rule of async programming is that **sync is simpler than async**. All the genius of Swift await/async does not invalidate this rule. Corollary: users who can use sync should use sync, and we must make it trivial to do so.

A user of an actor-based `AsyncDatabaseQueue` or `AsyncDatabasePool` who wants sync access would have to open an extra sync-enabled `DatabaseQueue` or `DatabasePool` on the side. But this goes against the [1st concurrency rule](https://github.com/groue/GRDB.swift/blob/master/README.md#concurrency): "Have a unique instance of DatabaseQueue or DatabasePool connected to any database file."

There are good reasons for this rule:
- Writes could throw SQLITE_BUSY errors.
- Database observation started on one connection would fail notifying changes performed by the other connection.

So, are actors a good fit for GRDB? Because of the reasons above, I think they are not. We must think about users before us, even if we are eager to use the new Swift async powers.

This is the first round of my ideas about this topic. What do you think?

**Reply by patskovn** [reply]
Well. Seems like we don't have any other opinions shared :)

> do we agree that Phase 1 could be implemented as extension methods in an application?

I assume that by "application" you mean any concrete iOS target? Yes, I thought about it too. Users may implement their own wrappers if they need them. But, in my opinion, its better to provide some "native" wrapper to users. I use GRDB since 2.0 and I didn't know that it have closure-based methods before I specifically search for new async wrappers, because I have always were using sync versions. So I think, library users **expect** that GRDB will have native wrappers like it have Rx/Combine versions.

> So I assume that actor-based database apis can only be async

Not really. Now you have `DatabaseQueue`, but you have additional `SerializedDatabase` which implements synchronous consecutive access to DB connection. You may have `DatabaseActor` which synchronizes all access and something like `SerializedDatabase` with sync connection which can be exposed as `nonisolated`. I understand, that `SerializedDatabase` synchronises its access to db by `DispatchQueue` again and wrapping `SerializedDatabase` to `DatabaseActor` will do nothing, because, well... you again have `DispatchQueue`-based synchronization.

Also, recently I have read article about performance of actors and looks like they are not so fast in comparison with `DispatchQueue` and locks. But it was synthetic tests. So all this actor-based synchronisations needs to be well tested. Does new implementation really will be more performant? I am not sure now.

**Reply by groue** [reply]
Hey, @Nekitosss :-)

> Well. Seems like we don't have any other opinions shared :)

Maybe I sounded too negative in my initial answer... But yes, people are shy. Welcome in, pals!

> Yes, I thought about it too. Users may implement their own wrappers if they need them. But, in my opinion, its better to provide some "native" wrapper to users.

All right, we agree.

I didn't tell why I asked this question.

If it can all be implemented as an extension, written in userland (not in the lib) of DatabaseReader/Writer or DatabaseQueue/Pool, then we have several advantages:

- Those apis can be provided in app, or in a third-party library written by some user: this means that *users are not blocked*. This is important in order to evaluate the urgency of shipping async GRDB apis. Yes, a "native", aka ready-made, library under the GRDB umbrella would make users' life even easier (right inside GRDB itself, or in a new companion library "AsyncGRDB"). But *we have time*, and this is important.
- If phase 1 apis should eventually be *replaced* by apis introduced in phase 2 and later, then we'd like to *remove*  phase 1 apis. Deprecation helps, but deprecated apis have to be supported. It would be even easier to remove them if they are not embedded in GRDB. We'd just say "stop using AsyncGRDB." Yes, my initial understanding of your OP was that phase 2+ apis would phase out the phase 1 apis.

Now you know why I'm happy that naive async wrappers *can* be implemented in userland without any change in the core lib.

I'll answer to your interesting actor feedback in another post 👍

**Reply by groue** [reply]
> [...] *users are not blocked*. This is important in order to evaluate the urgency of shipping async GRDB apis. Yes, a "native", aka ready-made, library under the GRDB umbrella would make users' life even easier (right inside GRDB itself, or in a new companion library "AsyncGRDB"). But *we have time*, and this is important.

Implicit in this sentence is the fact that GRDB has always grown this way: missing features are developed in applications first, according to my needs, the needs of my coworkers, the needs of users who are not shy, ask questions, send feature requests. This makes it possible to play with an idea in a few apps and different contexts, better understand the scope of the problem, and apply [the rule of 3](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) before the library apis are enlarged. It has never been an urgency in this lib to provide apis that *can be implemented in userland*. This method has allowed the lib to slowly grow over time and still address a lot of application needs.

**Reply by groue** [reply]
Regarding your comments about actors: I'm still wondering how we could provide database serialization with an unisolated actor, without using something like a serial dispatch queue or a lock, and thus ruining the benefits of actors and the Swift runtime.

**Reply by groue** [reply]
A recent though about the Phase 2/3 `AsyncDatabaseQueue` or `AsyncDatabasePool`:

Those new kinds of "connections" would be very different from the existing `DatabaseQueue` and `DatabasePool`:

- No dependency of DispatchQueue
- Fully based on genuine Swift async constructs, including actors
- Very probably async-only (no sync database accesses)

At first I was reluctant about removing support for sync database accesses (my "sync is simpler than async" paragraph in [this comment](https://github.com/groue/GRDB.swift/discussions/1065#discussioncomment-1383430)).

But if we design those new connections that that the user can bring back sync access by replacing `AsyncDatabaseQueue/Pool` with `DatabaseQueue/Pool`, with a **small** refactoring cost, then I don't see why we shouldn't explore this area.

**Reply by patskovn** [reply]
> without using something like a serial dispatch queue or a lock

No-no, sorry, I mean additional protection layer with lock. 
So my initial plan is

![Untitled-2021-10-13-1551-2](https://user-images.githubusercontent.com/17741730/137137695-58fc7506-ed5d-471d-9c8d-efd148ce5eba.png)

Also, DatabaseActor **expose `SerializedDatabase` pucblicly** like:
```swift 
nonisolated public func withBlocking(connector: DatabaseReader & DatabaseWriter)
```
or
```swift
nonisolated public var syncConnector: DatabaseReader & DatabaseWriter
```

My intuition tells me that additional synchronisation layer via locks is extremely cheap, considering the fact that 99% of calls will be already serialised by actor. Synchronous access is very useful for initizlization step, for example. When user wants to synchronously create database connection and perform initial setup synchronously.
But, as I told you above, I am not 100% sure about performance despite the fact that lock-based approach looks promising.

P.S. Sorry for long responses. I am busy with immigrating to another country. This took huge amount of my attention and time in this covid times.

**Reply by groue** [reply]
I do not understand yet, but I think it is because of the name `SerializedDatabase`. Maybe you use it as a new name and concept? It happens that there already exist [a class with this name](https://github.com/groue/GRDB.swift/blob/v5.12.0/GRDB/Core/SerializedDatabase.swift), which is internal and not intended for public use. It provides the foundations for high-level connectors like `DatabaseQueue` and `DatabasePool`, and is responsible for very low-level concurrency primitives and checks (DispatchQueue juggling and checks for transaction boundaries).

> Sorry for long responses.

That's totally OK. I'm very glad to discuss this delicate subject 👍

**Comment by groue** [comment]
**Which apis should get an async version?**

The "Phase 1" is making progress in the [dev/async](https://github.com/groue/GRDB.swift/tree/dev/async) branch. Those database access methods now have an `async` version, all based on lower level apis with completion blocks:

- `read`
- `unsafeRead`
- `write`
- `writeWithoutTransaction`
- `barrierWriteWithoutTransaction`
- `erase`
- `vacuum`
- `vacuum(into:)`

About the async version of `backup`, I'm waiting for #1075 by @mallman to make progress. I'm also wondering wether the backup should wait for an access to the source database first, or an access to the destination database first (who do we `await` for?). Advice is welcome.

I did not start thinking about how we can build an async sequence out of `ValueObservation`.

I'm even less sure how `DatabaseRegionObservation` should be turned into an async sequence. It publishes *database connections*, and thus requires, unlike `ValueObservation`, that callbacks are executed in a very specific DispatchQueue (GRDB fatal errors when a database connection is not used in a blessed DispatchQueue). To me this sounds like there are many opportunities for misuses, and that users will have great difficulties using it correctly (i.e. without crashing). I hope I'm wrong, and that we can _document_ a reliable way to use it.

Which other apis should become async?

**Reply by groue** [reply]
Migrations should become async.

EDIT: done in bc245c3f841a0e47a1d2b1f4717717a2c92ef19b

**Reply by StuClift** [reply]
Is this too naïve as a wrapper?

```swift
extension ValueObservation {
    func stream(
        in reader: DatabaseReader,
        scheduling scheduler: ValueObservationScheduler = .async(onQueue: .main)
    ) -> AsyncThrowingStream<Reducer.Value, Error> {
        AsyncThrowingStream { continuation in
            let cancellable = start(in: reader, scheduling: scheduler) {
                continuation.finish(throwing: $0)
            } onChange: {
                continuation.yield($0)
            }
            continuation.onTermination = { @Sendable _ in
                cancellable.cancel()
            }
        }
    }
}
```

**Reply by groue** [reply]
Thank you @StuClift! 🙂 

Do you know how this sentence from [the doc of `AsyncThrowingStream.init`](https://developer.apple.com/documentation/swift/asyncthrowingstream/3856746-init) should be understood? It happens that raw ValueObservation callbacks `onChange` and `onError` run on a single serial DispatchQueue and are strictly serialized. But is a single DispatchQueue a single "context" that avoids all risks of "out of order delivery"?

> The build closure passes in a Continuation which can be used in concurrent contexts. It is thread safe to send and finish; all calls to the continuation are serialized, **however calling this from multiple concurrent contexts could result in out of order delivery**.

**Reply by groue** [reply]
@StuClift, I don't quite understand why the user would like to provide an explicit `scheduler`. It looks like the `.immediate` scheduler, which is very useful in a synchronous context, is out of place in the async/await context. Do I miss anything?

**Reply by StuClift** [reply]
> Do you know how this sentence from the doc of AsyncThrowingStream.init should be understood? It happens that raw ValueObservation callbacks onChange and onError run on a single serial DispatchQueue and are strictly serialized. But is a single DispatchQueue a single "context" that avoids all risks of "out of order delivery"?

It is a little confusing because it kind of leads you into doubt as it seems like an obvious statement if you take it to mean it as you've stated; which I assume must be the case.

> @StuClift, I don't quite understand why the user would like to provide an explicit scheduler. It looks like the .immediate scheduler, which is very useful in a synchronous context, is out of place in the async/await context. Do I miss anything?

I added it to match the parameters available to `start` assuming this existed so the caller might have a reason to control the queue `onChange(_:)` handler was called from.

Without passing this through it wouldn't it be impossible to avoid hitting the main queue (even if only to call `yield(_:)` for each change?

Since this is a basic bridge to the GCD implementation; this might be a detail that matters to someone not awaiting these tasks through the `MainActor`.

**Reply by groue** [reply]
Thanks for your answers @StuClift. OK we'll keep the `scheduler` argument. All those async apis will be flagged as _experimental_ anyway, so that we can easily introduce breaking changes whenever we better understand the user needs and figure out better implementation strategies.

Practically speaking:

On the naming side, I may prefer `values` over `stream`, following [the example from Combine](https://developer.apple.com/documentation/combine/publisher/values-v7nz).

I could not figure out how to trigger the observation cancellation (`onTermination`) when I break the for loop:

```swift
for try await value in observation.values(in: dbQueue) {
    if someCondition {
        break // I expect this to cancel the observation eventually
    }
}
```

No matter how I tried, I'm currently unable to cancel the observation when the loop is broken. I have tried everything I could, including making very pessimistic hypothesis, such as considering the async runtime may _only_, and _eventually_, propagate cancellation when some parent Task, in which the loop is performed, is completed. But wrapping the broken loop in a Task, and waiting for the cancellation after the Task has been awaited yields no success.

Building the async sequence out of the ready-made Combine publisher didn't bring any support for cancellation either:

```swift
extension ValueObservation {
    public typealias Stream = Combine.AsyncThrowingPublisher<DatabasePublishers.Value<Reducer.Value>>
    
    func values(in reader: DatabaseReader) -> Stream {
        publisher(in: reader, scheduling: .async(onQueue: .main)).values
    }
}
```

I'm quite confused. I don't know how to turn a `break` into a cancellation. 😕 It's time to pick up a different topic for today.

**Reply by groue** [reply]
> I'm quite confused.

I asked on the Swift forums: https://forums.swift.org/t/asyncsequence-break-and-cancellation/52787

**Reply by groue** [reply]
OK, it's a [bug in the Swift 5.5 compiler](https://forums.swift.org/t/asyncsequence-break-and-cancellation/52787/5): `break` is not correctly handled. That's unfortunate for infinite sequences such as `ValueObservation` 🤷‍♂️

**Reply by groue** [reply]
> Is this too naïve as a wrapper?

Well, @StuClift, yes it is too naïve. I've learned it the hard way...

If you're interested, you may have a look at the current implementation: https://github.com/groue/GRDB.swift/blob/bbd7716d9ef841262ebd484bb0a58e453758fb17/GRDB/ValueObservation/ValueObservation.swift#L214-L315

Practically speaking,

We use an `AsyncThrowingStream`, as you suggest, because it buffers elements. We need to buffer elements because nothing guarantees that the user can loop over the database changes fast enough. And I do not want to write code that buffers elements in an async environment. This looks very hard. So `AsyncThrowingStream` it is.

Unfortunately, *I could never* have a raw AsyncThrowingStream handle Task cancellation and end-of-loop (`break`, early `return`, etc) correctly, which means cancel the ValueObservation.

So the stream is wrapped in a thin wrapper, `AsyncValueObservation` (which is just what `AsyncPublisher` is to `Publisher`), whose sole purpose is to introduce weak references where there need to be.

Now my goal is to test even more ways to interrupt an async loop, and make sure that the observation is properly cancelled:

- [x] immediate break
- [x] break after a while
- [x] sequence prefix
- [ ] Task cancellation
- [ ] etc.

Lesson learned: never assume that if one case is handled correctly, others are as well. Writing custom async sequences _that must release resources when they end_ is not easy. This is a _great_ opportunity to learn.

**Comment by groue** [comment]
Pull request for Phase 1: https://github.com/groue/GRDB.swift/pull/1079

Public announcement: [GRDB: Async/await edition is looking for testers](https://forums.swift.org/t/grdb-async-await-edition-is-looking-for-testers/52869)

**Reply by patskovn** [reply]
That's wonderful. Will test it in my personal project. I mostly use only reads and writes and does not rely on value observation, so my test scope won't be wide 👍🏻

**Comment by groue** [comment]
GRDB 5.17.0 is out, with the "phase 1" of async/await, as [framed](https://github.com/groue/GRDB.swift/discussions/1065#discussion-3589020) by @Nekitosss.

**Future evolutions will be driven by feedback on this initial release.**
